git/git.go

```diff
@@ -0,0 +1,145 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/ProtonMail/go-crypto/openpgp"
+)
+
+type Hash []byte
+
+// String returns the SHA1 Hash as a string.
+func (h Hash) String() string {
+	return string(h)
+}
+
+// Signature represents an entity which associates a person and a time
+// with a commit.
+type Signature struct {
+	Name  string
+	Email string
+	When  time.Time
+}
+
+// Commit contains all possible information about a Git commit.
+type Commit struct {
+	// Hash is the SHA1 hash of the commit.
+	Hash Hash
+	// Reference is the original reference of the commit, for example:
+	// 'refs/tags/foo'.
+	Reference string
+	// Author is the original author of the commit.
+	Author Signature
+	// Committer is the one performing the commit, might be different from
+	// Author.
+	Committer Signature
+	// Signature is the PGP signature of the commit.
+	Signature string
+	// Encoded is the encoded commit, without any signature.
+	Encoded []byte
+	// Message is the commit message, contains arbitrary text.
+	Message string
+}
+
+// String returns a string representation of the Commit, composed
+// out the last part of the Reference element, and/or Hash.
+// For example: 'tag-1/a0c14dc8580a23f79bc654faa79c4f62b46c2c22',
+// for a "tag-1" tag.
+func (c *Commit) String() string {
+	if short := strings.SplitAfterN(c.Reference, "/", 3); len(short) == 3 {
+		return fmt.Sprintf("%s/%s", short[2], c.Hash)
+	}
+	return fmt.Sprintf("HEAD/%s", c.Hash)
+}
+
+// Verify the Signature of the commit with the given key rings.
+// It returns the fingerprint of the key the signature was verified
+// with, or an error.
+func (c *Commit) Verify(keyRing ...string) (string, error) {
+	if c.Signature == "" {
+		return "", fmt.Errorf("commit does not have a PGP signature")
+	}
+
+	for _, r := range keyRing {
+		reader := strings.NewReader(r)
+		keyring, err := openpgp.ReadArmoredKeyRing(reader)
+		if err != nil {
+			return "", fmt.Errorf("failed to read armored key ring: %w", err)
+		}
+		signer, err := openpgp.CheckArmoredDetachedSignature(keyring, bytes.NewBuffer(c.Encoded), bytes.NewBufferString(c.Signature), nil)
+		if err == nil {
+			return fmt.Sprintf("%X", signer.PrimaryKey.Fingerprint[12:20]), nil
+		}
+	}
+	return "", fmt.Errorf("failed to verify commit with any of the given key rings")
+}
+
+// ShortMessage returns the first 50 characters of a commit subject.
+func (c *Commit) ShortMessage() string {
+	subject := strings.Split(c.Message, "\n")[0]
+	r := []rune(subject)
+	if len(r) > 50 {
+		return fmt.Sprintf("%s...", string(r[0:50]))
+	}
+	return subject
+}
+
+// GitClient defines all the methods that a git implementation must cover.
+// An instance of GitClient should have a 1:1 mapping with a Git repository.
+type GitClient interface {
```

Comment:
Wondering if it would be an idea to split this interface into multiple categories (e.g. read/write).

Reply:
+1 making this more composable would make it easier to test on the different use cases.

Reply:
ATM, the `GoGitClient` is stateful, i.e. it stores repository and related metadata. Seeing how projects like IAC and flux2 perform both read and write operations to the same repository, it might be tricky to design an API which splits this interface, since both would require access to the repository. In that case, we might need to make the clients stateless. @pjbgf @hiddeco

Reply:
@aryan9600 that is a good point, specially considering that some use cases do not require that statefulness - i.e. SC and I am assuming flux CLI as well.

By segregating this interface into smaller interfaces, we don't force the API consumers to implement everything when they only need a `Clone` operation, for example.

Another point is how we are planning to manage the release of lower level state that might need to be kept depending on the lower-level implementation (e.g. libgit2)? It would be great if that could be encapsulated, but I am not sure we can without making the implementation a lot less inefficient (i.e. creating a new repository instance for every single interface call).